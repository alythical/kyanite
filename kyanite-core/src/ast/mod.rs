pub mod init;
pub mod node;
pub mod span;
#[cfg(test)]
mod strip;

use crate::{
    parse::Parser,
    token::{Lexer, Token},
    PipelineError, Source,
};
use std::{fmt, rc::Rc};

#[derive(Debug)]
pub struct Ast {
    pub nodes: Vec<Decl>,
}

impl Ast {
    fn new(lexer: Lexer) -> Result<Self, PipelineError> {
        let errors = lexer.errors.len();
        if errors > 0 {
            return Err(PipelineError::LexError(errors));
        }
        let mut parser = Parser::new(lexer.source, lexer.tokens);
        match parser.parse() {
            Ok(nodes) => Ok(Self { nodes }),
            Err(errors) => Err(PipelineError::ParseError(errors.len())),
        }
    }
}

impl TryFrom<&Source> for Ast {
    type Error = PipelineError;

    fn try_from(value: &Source) -> Result<Self, Self::Error> {
        let lexer = Lexer::from(value);
        Self::new(lexer)
    }
}

#[derive(Debug, Clone)]
pub enum Decl {
    Record(Rc<node::RecordDecl>),
    Function(Rc<node::FuncDecl>),
    Constant(Rc<node::ConstantDecl>),
}

#[derive(Debug, Clone)]
pub enum Stmt {
    Var(Rc<node::VarDecl>),
    Assign(Rc<node::Assign>),
    Return(Rc<node::Return>),
    Expr(Expr),
    If(Rc<node::If>),
    While(Rc<node::While>),
}

#[derive(Debug, Clone)]
pub enum Expr {
    Call(Rc<node::Call>),
    Access(Rc<node::Access>),
    Binary(Rc<node::Binary>),
    Unary(Rc<node::Unary>),
    Ident(Rc<node::Ident>),
    Init(Rc<node::Init>),
    Str(&'static str, Token),
    Int(i64, Token),
    Float(f64, Token),
    Bool(bool, Token),
}

impl Expr {
    pub fn ident(&self) -> &node::Ident {
        if let Expr::Ident(ident) = self {
            ident
        } else {
            panic!("called `Expr::ident()` on a non-ident")
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Str,
    Int,
    Float,
    Bool,
    Void,
    UserDefined(String),
}

impl From<&Token> for Type {
    fn from(value: &Token) -> Self {
        match value.lexeme.expect("token should have lexeme") {
            "str" => Self::Str,
            "int" => Self::Int,
            "float" => Self::Float,
            "bool" => Self::Bool,
            "void" => Self::Void,
            name => Self::UserDefined(name.to_string()),
        }
    }
}

impl From<Option<&Token>> for Type {
    fn from(token: Option<&Token>) -> Self {
        match token {
            Some(token) => Self::from(token),
            None => Self::Void,
        }
    }
}

impl PartialEq<Type> for Option<Token> {
    fn eq(&self, other: &Type) -> bool {
        match self {
            Some(token) => Type::from(token) == *other,
            None => *other == Type::Void,
        }
    }
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Str => "str",
                Self::Int => "int",
                Self::Float => "float",
                Self::Bool => "bool",
                Self::Void => "void",
                Self::UserDefined(name) => name,
            }
        )
    }
}

impl Expr {
    pub fn ty(&self) -> Type {
        match self {
            Expr::Str(..) => Type::Str,
            Expr::Int(..) => Type::Int,
            Expr::Float(..) => Type::Float,
            Expr::Bool(..) => Type::Bool,
            Expr::Binary(binary) => binary.left.ty(),
            Expr::Unary(unary) => unary.expr.ty(),
            Expr::Call(call) => call.left.ty(),
            Expr::Ident(_) => unimplemented!(),
            Expr::Init(..) => unimplemented!(),
            Expr::Access(..) => unimplemented!(),
        }
    }
}

#[derive(Debug)]
pub struct Initializer {
    pub name: Token,
    pub expr: Expr,
}

impl Initializer {
    pub fn new(name: Token, expr: Expr) -> Self {
        Self { name, expr }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Param {
    pub name: Token,
    pub ty: Token,
}

impl Param {
    pub fn new(name: Token, ty: Token) -> Self {
        Self { name, ty }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Field {
    pub name: Token,
    pub ty: Token,
}

impl Field {
    pub fn new(name: Token, ty: Token) -> Self {
        Self { name, ty }
    }
}

macro_rules! assert_ast {
    ($($path:expr => $name:ident),*) => {
        #[cfg(test)]
        mod tests {
            use crate::{Source, ast::{self, strip::StripId}};

            $(
                #[test]
                fn $name() -> Result<(), Box<dyn std::error::Error>> {
                    let mut ast = ast::Ast::try_from(&Source::new($path)?)?;
                    ast.nodes.iter_mut().for_each(|node| node.strip_id());
                    insta::with_settings!({snapshot_path => "../../snapshots"}, {
                        insta::assert_debug_snapshot!(ast);
                    });
                    Ok(())
                }
            )*
        }
    };
}

assert_ast!(
    "test-cases/hello.kya" => hello_world,
    "test-cases/expr.kya" => expr,
    "test-cases/calls.kya" => calls,
    "test-cases/empty.kya" => empty,
    "test-cases/access.kya" => access,
    "test-cases/mixed.kya" => mixed
);
