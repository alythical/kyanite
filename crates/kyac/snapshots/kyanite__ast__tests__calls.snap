---
source: kyanite-core/src/ast/mod.rs
expression: ast
---
Ast {
    nodes: [
        Function(
            FuncDecl {
                name: Token {
                    kind: Identifier,
                    lexeme: Some(
                        "foo",
                    ),
                    span: Span {
                        line: 1,
                        column: 5,
                        length: 3,
                    },
                },
                params: [
                    Param {
                        name: Token {
                            kind: Identifier,
                            lexeme: Some(
                                "bar",
                            ),
                            span: Span {
                                line: 1,
                                column: 9,
                                length: 3,
                            },
                        },
                        ty: Token {
                            kind: Identifier,
                            lexeme: Some(
                                "int",
                            ),
                            span: Span {
                                line: 1,
                                column: 14,
                                length: 3,
                            },
                        },
                    },
                ],
                ty: Some(
                    Token {
                        kind: Identifier,
                        lexeme: Some(
                            "int",
                        ),
                        span: Span {
                            line: 1,
                            column: 20,
                            length: 3,
                        },
                    },
                ),
                body: [
                    Return(
                        Return {
                            expr: Ident(
                                Ident {
                                    name: Token {
                                        kind: Identifier,
                                        lexeme: Some(
                                            "bar",
                                        ),
                                        span: Span {
                                            line: 2,
                                            column: 11,
                                            length: 3,
                                        },
                                    },
                                },
                            ),
                            keyword: Token {
                                kind: Return,
                                lexeme: None,
                                span: Span {
                                    line: 2,
                                    column: 4,
                                    length: 6,
                                },
                            },
                        },
                    ),
                ],
                external: false,
                id: 0,
            },
        ),
        Function(
            FuncDecl {
                name: Token {
                    kind: Identifier,
                    lexeme: Some(
                        "bar",
                    ),
                    span: Span {
                        line: 5,
                        column: 5,
                        length: 3,
                    },
                },
                params: [],
                ty: Some(
                    Token {
                        kind: Identifier,
                        lexeme: Some(
                            "float",
                        ),
                        span: Span {
                            line: 5,
                            column: 12,
                            length: 5,
                        },
                    },
                ),
                body: [
                    Return(
                        Return {
                            expr: Float(
                                42.0,
                                Token {
                                    kind: Literal,
                                    lexeme: Some(
                                        "42.",
                                    ),
                                    span: Span {
                                        line: 6,
                                        column: 11,
                                        length: 3,
                                    },
                                },
                            ),
                            keyword: Token {
                                kind: Return,
                                lexeme: None,
                                span: Span {
                                    line: 6,
                                    column: 4,
                                    length: 6,
                                },
                            },
                        },
                    ),
                ],
                external: false,
                id: 0,
            },
        ),
        Function(
            FuncDecl {
                name: Token {
                    kind: Identifier,
                    lexeme: Some(
                        "helloWorld",
                    ),
                    span: Span {
                        line: 9,
                        column: 5,
                        length: 10,
                    },
                },
                params: [],
                ty: None,
                body: [
                    Expr(
                        Call(
                            Call {
                                left: Ident(
                                    Ident {
                                        name: Token {
                                            kind: Identifier,
                                            lexeme: Some(
                                                "println",
                                            ),
                                            span: Span {
                                                line: 10,
                                                column: 4,
                                                length: 7,
                                            },
                                        },
                                    },
                                ),
                                args: [
                                    Str(
                                        "\"Hello world!\"",
                                        Token {
                                            kind: Literal,
                                            lexeme: Some(
                                                "\"Hello world!\"",
                                            ),
                                            span: Span {
                                                line: 10,
                                                column: 12,
                                                length: 14,
                                            },
                                        },
                                    ),
                                ],
                                parens: (
                                    Token {
                                        kind: LeftParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 10,
                                            column: 11,
                                            length: 1,
                                        },
                                    },
                                    Token {
                                        kind: RightParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 10,
                                            column: 26,
                                            length: 1,
                                        },
                                    },
                                ),
                                delimiters: [],
                            },
                        ),
                    ),
                ],
                external: false,
                id: 0,
            },
        ),
        Function(
            FuncDecl {
                name: Token {
                    kind: Identifier,
                    lexeme: Some(
                        "main",
                    ),
                    span: Span {
                        line: 13,
                        column: 5,
                        length: 4,
                    },
                },
                params: [],
                ty: None,
                body: [
                    Expr(
                        Call(
                            Call {
                                left: Ident(
                                    Ident {
                                        name: Token {
                                            kind: Identifier,
                                            lexeme: Some(
                                                "helloWorld",
                                            ),
                                            span: Span {
                                                line: 14,
                                                column: 4,
                                                length: 10,
                                            },
                                        },
                                    },
                                ),
                                args: [],
                                parens: (
                                    Token {
                                        kind: LeftParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 14,
                                            column: 14,
                                            length: 1,
                                        },
                                    },
                                    Token {
                                        kind: RightParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 14,
                                            column: 15,
                                            length: 1,
                                        },
                                    },
                                ),
                                delimiters: [],
                            },
                        ),
                    ),
                    Expr(
                        Call(
                            Call {
                                left: Ident(
                                    Ident {
                                        name: Token {
                                            kind: Identifier,
                                            lexeme: Some(
                                                "println",
                                            ),
                                            span: Span {
                                                line: 15,
                                                column: 4,
                                                length: 7,
                                            },
                                        },
                                    },
                                ),
                                args: [
                                    Call(
                                        Call {
                                            left: Ident(
                                                Ident {
                                                    name: Token {
                                                        kind: Identifier,
                                                        lexeme: Some(
                                                            "foo",
                                                        ),
                                                        span: Span {
                                                            line: 15,
                                                            column: 12,
                                                            length: 3,
                                                        },
                                                    },
                                                },
                                            ),
                                            args: [
                                                Int(
                                                    72,
                                                    Token {
                                                        kind: Literal,
                                                        lexeme: Some(
                                                            "72",
                                                        ),
                                                        span: Span {
                                                            line: 15,
                                                            column: 16,
                                                            length: 2,
                                                        },
                                                    },
                                                ),
                                            ],
                                            parens: (
                                                Token {
                                                    kind: LeftParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 15,
                                                        column: 15,
                                                        length: 1,
                                                    },
                                                },
                                                Token {
                                                    kind: RightParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 15,
                                                        column: 18,
                                                        length: 1,
                                                    },
                                                },
                                            ),
                                            delimiters: [],
                                        },
                                    ),
                                ],
                                parens: (
                                    Token {
                                        kind: LeftParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 15,
                                            column: 11,
                                            length: 1,
                                        },
                                    },
                                    Token {
                                        kind: RightParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 15,
                                            column: 19,
                                            length: 1,
                                        },
                                    },
                                ),
                                delimiters: [],
                            },
                        ),
                    ),
                    Expr(
                        Call(
                            Call {
                                left: Ident(
                                    Ident {
                                        name: Token {
                                            kind: Identifier,
                                            lexeme: Some(
                                                "println",
                                            ),
                                            span: Span {
                                                line: 16,
                                                column: 4,
                                                length: 7,
                                            },
                                        },
                                    },
                                ),
                                args: [
                                    Call(
                                        Call {
                                            left: Ident(
                                                Ident {
                                                    name: Token {
                                                        kind: Identifier,
                                                        lexeme: Some(
                                                            "bar",
                                                        ),
                                                        span: Span {
                                                            line: 16,
                                                            column: 12,
                                                            length: 3,
                                                        },
                                                    },
                                                },
                                            ),
                                            args: [],
                                            parens: (
                                                Token {
                                                    kind: LeftParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 16,
                                                        column: 15,
                                                        length: 1,
                                                    },
                                                },
                                                Token {
                                                    kind: RightParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 16,
                                                        column: 16,
                                                        length: 1,
                                                    },
                                                },
                                            ),
                                            delimiters: [],
                                        },
                                    ),
                                ],
                                parens: (
                                    Token {
                                        kind: LeftParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 16,
                                            column: 11,
                                            length: 1,
                                        },
                                    },
                                    Token {
                                        kind: RightParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 16,
                                            column: 17,
                                            length: 1,
                                        },
                                    },
                                ),
                                delimiters: [],
                            },
                        ),
                    ),
                    Expr(
                        Call(
                            Call {
                                left: Ident(
                                    Ident {
                                        name: Token {
                                            kind: Identifier,
                                            lexeme: Some(
                                                "println",
                                            ),
                                            span: Span {
                                                line: 17,
                                                column: 4,
                                                length: 7,
                                            },
                                        },
                                    },
                                ),
                                args: [
                                    Call(
                                        Call {
                                            left: Ident(
                                                Ident {
                                                    name: Token {
                                                        kind: Identifier,
                                                        lexeme: Some(
                                                            "max",
                                                        ),
                                                        span: Span {
                                                            line: 17,
                                                            column: 12,
                                                            length: 3,
                                                        },
                                                    },
                                                },
                                            ),
                                            args: [
                                                Int(
                                                    10,
                                                    Token {
                                                        kind: Literal,
                                                        lexeme: Some(
                                                            "10",
                                                        ),
                                                        span: Span {
                                                            line: 17,
                                                            column: 16,
                                                            length: 2,
                                                        },
                                                    },
                                                ),
                                                Int(
                                                    300,
                                                    Token {
                                                        kind: Literal,
                                                        lexeme: Some(
                                                            "300",
                                                        ),
                                                        span: Span {
                                                            line: 17,
                                                            column: 20,
                                                            length: 3,
                                                        },
                                                    },
                                                ),
                                            ],
                                            parens: (
                                                Token {
                                                    kind: LeftParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 17,
                                                        column: 15,
                                                        length: 1,
                                                    },
                                                },
                                                Token {
                                                    kind: RightParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 17,
                                                        column: 23,
                                                        length: 1,
                                                    },
                                                },
                                            ),
                                            delimiters: [
                                                Token {
                                                    kind: Comma,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 17,
                                                        column: 18,
                                                        length: 1,
                                                    },
                                                },
                                            ],
                                        },
                                    ),
                                ],
                                parens: (
                                    Token {
                                        kind: LeftParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 17,
                                            column: 11,
                                            length: 1,
                                        },
                                    },
                                    Token {
                                        kind: RightParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 17,
                                            column: 24,
                                            length: 1,
                                        },
                                    },
                                ),
                                delimiters: [],
                            },
                        ),
                    ),
                    Expr(
                        Call(
                            Call {
                                left: Ident(
                                    Ident {
                                        name: Token {
                                            kind: Identifier,
                                            lexeme: Some(
                                                "println",
                                            ),
                                            span: Span {
                                                line: 18,
                                                column: 4,
                                                length: 7,
                                            },
                                        },
                                    },
                                ),
                                args: [
                                    Call(
                                        Call {
                                            left: Ident(
                                                Ident {
                                                    name: Token {
                                                        kind: Identifier,
                                                        lexeme: Some(
                                                            "min",
                                                        ),
                                                        span: Span {
                                                            line: 18,
                                                            column: 12,
                                                            length: 3,
                                                        },
                                                    },
                                                },
                                            ),
                                            args: [
                                                Float(
                                                    10.0,
                                                    Token {
                                                        kind: Literal,
                                                        lexeme: Some(
                                                            "10.",
                                                        ),
                                                        span: Span {
                                                            line: 18,
                                                            column: 16,
                                                            length: 3,
                                                        },
                                                    },
                                                ),
                                                Float(
                                                    300.0,
                                                    Token {
                                                        kind: Literal,
                                                        lexeme: Some(
                                                            "300.",
                                                        ),
                                                        span: Span {
                                                            line: 18,
                                                            column: 21,
                                                            length: 4,
                                                        },
                                                    },
                                                ),
                                            ],
                                            parens: (
                                                Token {
                                                    kind: LeftParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 18,
                                                        column: 15,
                                                        length: 1,
                                                    },
                                                },
                                                Token {
                                                    kind: RightParen,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 18,
                                                        column: 25,
                                                        length: 1,
                                                    },
                                                },
                                            ),
                                            delimiters: [
                                                Token {
                                                    kind: Comma,
                                                    lexeme: None,
                                                    span: Span {
                                                        line: 18,
                                                        column: 19,
                                                        length: 1,
                                                    },
                                                },
                                            ],
                                        },
                                    ),
                                ],
                                parens: (
                                    Token {
                                        kind: LeftParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 18,
                                            column: 11,
                                            length: 1,
                                        },
                                    },
                                    Token {
                                        kind: RightParen,
                                        lexeme: None,
                                        span: Span {
                                            line: 18,
                                            column: 26,
                                            length: 1,
                                        },
                                    },
                                ),
                                delimiters: [],
                            },
                        ),
                    ),
                ],
                external: false,
                id: 0,
            },
        ),
    ],
}
